### PR TITLE
ROX-21178: Add buffer in the runtime detection path

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

Runtime detection (process indicators and network flows) are now sent to internal queues where the events are buffered after enhancement. The detector pulls from these internal queues and sends these events to the runtime detector.

- If the FF `ROX_CAPTURE_INTERMEDIATE_EVENTS` is enabled and Sensor is offline, these internal queues will be paused. This means no events can be pulled from them.
- If the queues are full, any new event will be dropped. This prevents Sensor from OOMing if we are offline.
- Metrics for the queues' size and number of events dropped have been added.
- The sizes are configurable with `ROX_SENSOR_DETECTOR_PROCESS_INDICATOR_BUFFER_SIZE` and `ROX_SENSOR_DETECTOR_NETWORK_FLOW_BUFFER_SIZE`

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

- [x] The pausable queue is covered with unit tests.
- [x] Manual testing:
  - Deploy ACS
  - Create a system policy that checks for the execution of a specific command (e.g. `curl`)
  - Break the connection between Central and Sensor
  - Create a pod and run some (`curl`) commands: `kubectl -n test run talk --rm -it --image alpine -- /bin/sh`
  - Re-establish the connection between Central and Sensor
  - Observe the violations that appear in the UI

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by vikin91 — COMMENTED** [review]
I am not super familiar with inner workings of detector, so I did only a partial review.

I find the composition of queues a bit strange. You have a `queue.Queue`(detector/queue) that contains a `queue.PausableQueue` that has `internalQueue` of type `queue.Queue` (but this time `pkg/queue`).  Not saying this is wrong, but navigating that is a bit difficult. (Unfortunately, for now, I do not have a clear idea how to improve that).

File: pkg/queue/pausable_queue.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+			val, aggregated := fn(element, item)
+			if aggregated {
+				e.Value = val
+				wasAggregated = true
+				break
+			}
+		}
+		if wasAggregated {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply pkg/queue/pausable_queue.go]: I've removed the aggregator logic in the last commit but your comment made sense 🙂

File: pkg/queue/pausable_queue.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+// Pull will pull an item from the queue. If the queue is empty or paused, the default value of T will be returned.
+// Note that his does not wait for items to be available in the queue, use PullBlocking instead.
+func (q *PausableQueue[T]) Pull() T {
+	var ret T
+	if !q.isRunning.IsDone() {
+		return ret
+	}
+
+	return q.internalQueue.Pull()
+}
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply pkg/queue/pausable_queue.go]: I was just mimicking the interface of the other pkg/queue, but I think your comment makes a lot of sense.

**Comment by lvalerom** [comment]
> I find the composition of queues a bit strange. You have a queue.Queue(detector/queue) that contains a queue.PausableQueue that has internalQueue of type queue.Queue (but this time pkg/queue). Not saying this is wrong, but navigating that is a bit difficult. (Unfortunately, for now, I do not have a clear idea how to improve that).

The idea was to add new functionality (pausable/aggregator) to the pkg/queue. Since we don't need the aggregator logic anymore I've simplified the code a bit. Now the detector/queue holds the pause/resume logic and makes use of pkg/queue to hold the elements and metrics.

File: sensor/common/detector/metrics/metrics.go [hunk 3]
```diff
@@ -88,6 +88,38 @@ var (
...
+		Subsystem: metrics.SensorSubsystem.String(),
+		Name:      "detector_network_flow_buffer_size",
+		Help:      "A counter that tracks the size of the detection network flow buffer",
+	}, []string{"Operation"})
+
+	// DetectorProcessIndicatorDroppedCount keeps track of the number of process indicators dropped in the detector.
+	DetectorProcessIndicatorDroppedCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.SensorSubsystem.String(),
+		Name:      "detector_process_indicator_dropper_counter",
```
Comment (vikin91) [inline sensor/common/detector/metrics/metrics.go:112]: According to https://prometheus.io/docs/practices/naming/#metric-names, we should rather name it 
```suggestion
		Name:      "detector_process_indicators_dropped_total",
```

File: sensor/common/detector/metrics/metrics.go [hunk 4]
```diff
@@ -88,6 +88,38 @@ var (
...
+		Help:      "A counter of the total number of process indicators that were dropped if the detector buffer was full",
+	})
+
+	// DetectorNetworkFlowDroppedCount keeps track of the number of network flows dropped in the detector.
+	DetectorNetworkFlowDroppedCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.SensorSubsystem.String(),
+		Name:      "detector_network_flow_dropper_counter",
```
Comment (vikin91) [inline sensor/common/detector/metrics/metrics.go:120]:
```suggestion
		Name:      "detector_network_flows_dropped_total",
```

**Review by vikin91 — APPROVED** [review]
Understanding the detector code requires much more experience with that part than I have, but to the best of my knowledge it looks good! 
I went over the entire change and was able to find only minor issues with metrics naming (I suggest to correct them before merging).

File: sensor/common/detector/metrics/metrics.go (hunk 3 above)
Comment (lvalerom) [reply sensor/common/detector/metrics/metrics.go]: Done

File: sensor/common/detector/metrics/metrics.go (hunk 4 above)
Comment (lvalerom) [reply sensor/common/detector/metrics/metrics.go]: Done

**Pull request merged**